gctscript/vm/vm.go

```diff
@@ -0,0 +1,202 @@
+	log "github.com/thrasher-corp/gocryptotrader/logger"
+)
+
+func newVM() *VM {
+	newUUID, err := uuid.NewV4()
+	if err != nil {
+		log.Error(log.GCTScriptMgr, Error{
+			Action: "New -> UUID",
+			Cause:  err,
+		})
+		return nil
+	}
+
+	if GCTScriptConfig.Verbose {
+		log.Debugln(log.GCTScriptMgr, "New GCTScript VM created")
+	}
+
+	audit.Event(newUUID.String(), AuditEventName, "virtual machine created")
+
+	return &VM{
+		ID:     newUUID,
+		Script: pool.Get().(*script.Script),
+	}
+}
+
+// Load parses and creates a new instance of tengo script vm
+func (vm *VM) Load(file string) error {
+	if !GCTScriptConfig.Enabled {
+		return &Error{
+			Action: "Load",
+			Cause:  ErrScriptingDisabled,
+		}
+	}
+
+	if GCTScriptConfig.Verbose {
+		log.Debugf(log.GCTScriptMgr, "Loading script: %v", file)
+	}
+
+	f, err := os.Open(file)
+	if err != nil {
+		return &Error{
+			Action: "Load -> Open",
+			Script: file,
+			Cause:  err,
+		}
+	}
+
+	code, err := ioutil.ReadAll(f)
+	if err != nil {
+		return &Error{
+			Action: "Load -> Read",
+			Script: file,
+			Cause:  err,
+		}
+	}
+
+	vm.file = f.Name()
+	vm.Name = vm.shortName(file)
+	vm.Script = script.New(code)
+	vm.Script.SetImports(loader.GetModuleMap())
+
+	if GCTScriptConfig.AllowImports {
+		if GCTScriptConfig.Verbose {
+			log.Debugf(log.GCTScriptMgr, "file imports enabled for vm: %v", vm.ID)
+		}
+		vm.Script.EnableFileImport(true)
+	}
+	audit.Event(vm.ID.String(), AuditEventName, "Script loaded")
+	return nil
+}
+
+// Compile compiles to byte code loaded copy of vm script
+func (vm *VM) Compile() (err error) {
+	vm.Compiled = new(script.Compiled)
+	vm.Compiled, err = vm.Script.Compile()
+	return
+}
+
+// Run runs byte code
+func (vm *VM) Run() (err error) {
+	audit.Event(vm.ID.String(), AuditEventName, "Script executed")
+	return vm.Compiled.Run()
+}
+
+// RunCtx runs compiled byte code with context.Context support.
+func (vm *VM) RunCtx() (err error) {
+	if vm.ctx == nil {
+		vm.ctx = context.Background()
+	}
+
+	ct, cancel := context.WithTimeout(vm.ctx, GCTScriptConfig.ScriptTimeout)
+	defer cancel()
+
+	audit.Event(vm.ID.String(), AuditEventName, "Script executed")
+
+	err = vm.Compiled.RunContext(ct)
+	if err != nil {
+		return Error{
+			Action: "RunCtx()",
+			Cause:  err,
+		}
+	}
+	return
+}
+
+// CompileAndRun Compile and Run script with support for task running
+func (vm *VM) CompileAndRun() {
+	err := vm.Compile()
+	if err != nil {
+		log.Error(log.GCTScriptMgr, err)
+		err = RemoveVM(vm.ID)
+		if err != nil {
+			log.Error(log.GCTScriptMgr, err)
+		}
+		return
+	}
+
+	if GCTScriptConfig.Verbose {
+		log.Debugf(log.GCTScriptMgr, "Running script: %v", vm.ID)
+	}
+
+	err = vm.RunCtx()
+	if err != nil {
+		log.Error(log.GCTScriptMgr, err)
+		err = RemoveVM(vm.ID)
+		if err != nil {
+			log.Error(log.GCTScriptMgr, err)
+		}
+		return
+	}
+
+	if vm.Compiled.Get("timer").String() != "" {
+		vm.T, err = time.ParseDuration(vm.Compiled.Get("timer").String())
```

Comment:
Additional execution flow & control is something I plan on adding

I also want to support a clean up function in the scripts that when gct is shutdown it can call and execute 

For the second party i have started to add audit logging into common called features but it sounds like you might think this is better off in its own table as well?

Reply:
Nice :) Will a script be able to launch another script too?

I think audit logging will be good to say "something has happened", but a dedicated table that would have more details in their own columns would be good for tracking.

Reply:
They can't but it can be added (scripts can import from other scripts at the moment)

I would need to add additional checks into for it to stop some crazy looping not sure if it will make it into this PR

Reply:
Should have added by "import" its pretty basic they can import functions from other scripts as a way to keep them clean 
you might have a dosomething.gct that is imported by main.gct dosomething has an exported function that returns 1+1

Reply:
I'm not really sure I'd want script launching from scripts anyway. Any repeated complexities should be turned into functions within gct rather than calling other scripts to prevent loops like you said.

Reply:
Is execution flow and control still being added to this PR?

Reply:
full execution flow and control has not made it into this PR due to the size of it already

I do still want to add at least:

1) exit on condition
2) pause on condition

a very very very basic example of exit control could be setting timer to 0 at runtime

```go
fmt := import("fmt")
os := import("os")

timer := "5s"

exit := func() {
    timer = 0
}

load := func() {
   for x := 0 ; x < 20; x++ {
      fmt.printf("Hello %v", x)
   }
   exit()
}

load()
```

But this is not idea and i do plan on improving support for better control greatly

Reply:
That's fair. This is a big PR already. I'll click resolve since you've addressed everything else 👍